Modules/posixmodule.c

```diff
@@ -9115,16 +9115,104 @@ os_setpgrp_impl(PyObject *module)
 #ifdef HAVE_GETPPID
 
 #ifdef MS_WINDOWS
-#include <processsnapshot.h>
+#include <winternl.h>
+
+// The structure definition in winternl.h may be incomplete.
+// This structure is the full version from the MSDN documentation.
+typedef struct _PROCESS_BASIC_INFORMATION_FULL {
+    NTSTATUS ExitStatus;
+    PVOID PebBaseAddress;
+    ULONG_PTR AffinityMask;
+    LONG BasePriority;
+    ULONG_PTR UniqueProcessId;
+    ULONG_PTR InheritedFromUniqueProcessId;
+} PROCESS_BASIC_INFORMATION_FULL;
+
+typedef NTSTATUS (NTAPI *PNT_QUERY_INFORMATION_PROCESS) (
+    IN    HANDLE           ProcessHandle,
+    IN    PROCESSINFOCLASS ProcessInformationClass,
+    OUT   PVOID            ProcessInformation,
+    IN    ULONG            ProcessInformationLength,
+    OUT   PULONG           ReturnLength OPTIONAL);
+
+// This function returns the process ID of the parent process.
+// Returns 0 on failure.
+static ULONG
+win32_getppid_fast(void)
+{
+    NTSTATUS Status;
+    HMODULE Ntdll;
+    PNT_QUERY_INFORMATION_PROCESS pNtQueryInformationProcess;
+    PROCESS_BASIC_INFORMATION_FULL BasicInformation;
+    static ULONG CachedParentProcessId = 0;
+
+    if (CachedParentProcessId) {
+        // No need to query the kernel again.
+        return CachedParentProcessId;
+    }
+
+    Ntdll = GetModuleHandleW(L"ntdll.dll");
+    if (!Ntdll) {
+        return 0;
+    }
+
+    pNtQueryInformationProcess = (PNT_QUERY_INFORMATION_PROCESS) GetProcAddress(Ntdll, "NtQueryInformationProcess");
+    if (!pNtQueryInformationProcess) {
+        return 0;
+    }
+
+    Status = pNtQueryInformationProcess(
+        GetCurrentProcess(),
+        ProcessBasicInformation,
+        &BasicInformation,
+        sizeof(BasicInformation),
+        NULL);
+
+    if (!NT_SUCCESS(Status)) {
+        return 0;
+    }
+
+    //
+    // Perform sanity check on the parent process ID we received from NtQueryInformationProcess.
+    // The check covers values which exceed the 32-bit range (if running on x64) as well as
+    // zero and (ULONG) -1.
+    //
+
+    if (BasicInformation.InheritedFromUniqueProcessId == 0 ||
+        BasicInformation.InheritedFromUniqueProcessId >= ULONG_MAX) {
+
```

Comment:
As the [PEP7 said](https://peps.python.org/pep-0007/#code-lay-out), the brace should start in a new line if break a long expression at a binary operator:

```suggestion
        BasicInformation.InheritedFromUniqueProcessId >= ULONG_MAX)
    {
```